python/pyspark/sql/types.py

```diff
@@ -1694,6 +1694,21 @@ def from_arrow_schema(arrow_schema):
          for field in arrow_schema])
 
 
+def _correct_date_of_dataframe_from_arrow(pdf, schema):
+    """ Correct date type value to use datetime.date.
+
+    Pandas DataFrame created from PyArrow uses datetime64[ns] for date type values, but we should
+    use datetime.date to keep backward compatibility.
```

Comment:
Shall we say like to match it with when Arrow optimization is disabled?

Reply:
Maybe we don't need to say about backward compatibility here. I'll update it.